src/librustc_resolve/late/diagnostics.rs

```diff
@@ -353,10 +353,252 @@ impl<'a> LateResolutionVisitor<'a, '_, '_> {
                 }
                 _ => {}
             }
+            if ns == TypeNS && path.len() == 1 {
+                // Case-insensitive test against most libstd struct names
+                // as another fallback
+                match res {
+                    Some(Res::ToolMod) => {}
+                    _ => {
+                        let structs = self.get_case_insensitive_libstd_structs_matches(
+                            &ident.name.to_ident_string(),
+                        );
+                        if structs.len() == 1 {
+                            err.span_suggestion(
+                                ident_span,
+                                &format!("found a std struct with a similar name"),
+                                format!("{}", structs[0]),
+                                Applicability::MaybeIncorrect,
+                            );
+                        } else if structs.len() > 1 {
+                            let mut struct_suggestions = Vec::new();
+                            let message = "found std structs with similar names:";
+                            for a_struct in structs.iter() {
+                                struct_suggestions.push(format!("{}", a_struct));
+                            }
+                            err.span_suggestions(
+                                ident_span,
+                                message,
+                                struct_suggestions.into_iter(),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
+                    }
+                }
+            }
         }
+
         (err, candidates)
     }
 
+    /// Get a case-insensitive match with standard library
+    /// structs that are *not imported* in the prelude.
+    /// This is used for type checking diagnostics in cases when
+    /// the type is not in scope and the name includes case
+    /// misspelling (e.g., `Hashmap`, not `HashMap`).
+    fn get_case_insensitive_libstd_structs_matches(&self, needle: &str) -> Vec<String> {
```

Comment:
I think we should take a different approach - I'd rather we avoid hard-coding `std` type names in the compiler (the compiler shouldn't need to be updated when `std` is changed for this diagnostic to continue to work) and I think we can make this more general, so it works for more than just `std`.

---

As you note in the PR description, we do Levenshtein matching on things that are already in-scope and suggest fixes in that case. We also already have suggestions for things that aren't in scope but where the name matches exactly.

The snippet below is where the compiler decides to make "did you mean to import this" suggestions:

https://github.com/rust-lang/rust/blob/50c0192c64241d723066add22c53d472e2b9cba9/src/librustc_resolve/diagnostics.rs#L660-L693

On line 660 is where it checks for an exact match - I'd experiment with doing a exact match *or* checking for a Levenshtein distance from that name that's under some threshold (look into `lev_distance` and `find_best_match_for_name` to see what can be re-used). I'd hope that would detect cases like this where we can import a name that's very similar. Alternatively, we could just change the comparison to compare the names without considering the case (as I understand it, that would be equivalent to what you're doing here?).

What we might also want to do is thread this information through in some way - e.g. changing `ImportSuggestion` to be an enum with `Exact` and `Almost` (you can come up with a better name than this) variants so that we can make the diagnostic separate these two types of suggestions:

```
error[E0412]: cannot find type `Hashmap` in this scope
 --> src/main.rs:2:8
  |
2 |     m: Hashmap<String, ()>,
  |        ^^^^^^^ not found in this scope
  |
help: consider importing one of these items
  |
1 | use some::made:up::module::Hashmap;
  |
help: consider importing one of these similarly named items
  |
1 | use std::collections::HashMap;
  |
```

This should be easier to maintain because we don't have to keep a list of types like you've got here, and it'd work for non-`std` types too.

Reply:
Tyvm David!  I will look into it this week.  I really appreciate the review and will be in touch when I have updates.

Reply:
Small nitpick on wording: `help: consider importing one of these similarly named items` → `help: you might have meant to use {one of }the following similarly named item{s}`.

Reply:
@chrissimpkins How are you getting with this? Feel free to open a work-in-progress PR if you’re unsure how to proceed or want some early feedback!

Reply:
Sorry for the delay! Busy week last week. I will get back to this during the week and let you know if I run into any issues.